subxt/src/backend/mod.rs

```diff
@@ -325,9 +325,395 @@ pub enum TransactionStatus<Hash> {
 
 /// A response from calls like [`Backend::storage_fetch_values`] or
 /// [`Backend::storage_fetch_descendant_values`].
+#[cfg_attr(test, derive(serde::Serialize, Clone, PartialEq, Debug))]
 pub struct StorageResponse {
     /// The key.
     pub key: Vec<u8>,
     /// The associated value.
     pub value: Vec<u8>,
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    mod legacy {
+        use std::{
+            collections::{HashMap, VecDeque},
+            sync::Arc,
+        };
+        use tokio::sync::{mpsc, Mutex};
+
+        use super::rpc::{RpcClient, RpcClientT};
+        use crate::backend::rpc::RawRpcSubscription;
+        use crate::backend::BackendExt;
+        use crate::{
+            backend::{
+                legacy::rpc_methods::Bytes, legacy::rpc_methods::RuntimeVersion,
+                legacy::LegacyBackend, StorageResponse,
+            },
+            error::RpcError,
+        };
+        use futures::StreamExt;
+        use serde::Serialize;
+        use serde_json::value::RawValue;
+        use subxt_core::{config::DefaultExtrinsicParams, Config};
+        type RpcResult<T> = Result<T, RpcError>;
+        type Item = RpcResult<String>;
+
+        struct StorageEntry {
+            items: HashMap<Vec<u8>, VecDeque<Item>>,
+        }
+
+        trait Storage {
+            fn new() -> Self;
+            fn push<I: Serialize>(&mut self, key: Vec<u8>, item: RpcResult<I>);
+
+            fn pop(&mut self, key: Vec<u8>) -> Item;
+        }
+
+        impl Storage for StorageEntry {
+            fn new() -> Self {
+                StorageEntry {
+                    items: HashMap::new(),
+                }
+            }
+            fn push<I: Serialize>(&mut self, key: Vec<u8>, item: RpcResult<I>) {
+                let item = item.map(|x| serde_json::to_string(&x).unwrap());
+                match self.items.entry(key) {
+                    std::collections::hash_map::Entry::Occupied(v) => v.into_mut().push_back(item),
+                    std::collections::hash_map::Entry::Vacant(e) => {
+                        e.insert(VecDeque::from([item]));
+                    }
+                }
+            }
+
+            fn pop(&mut self, key: Vec<u8>) -> Item {
+                self.items.get_mut(&key).unwrap().pop_front().unwrap()
+            }
+        }
+
+        struct Data<R: Storage> {
+            request: R,
+            subscription: mpsc::Receiver<RpcResult<Vec<Item>>>,
+        }
+
+        struct MockRpcClientStorage {
+            data: Arc<Mutex<Data<StorageEntry>>>,
+        }
+
+        impl RpcClientT for MockRpcClientStorage {
+            fn request_raw<'a>(
+                &'a self,
+                method: &'a str,
+                params: Option<Box<serde_json::value::RawValue>>,
+            ) -> super::rpc::RawRpcFuture<'a, Box<serde_json::value::RawValue>> {
+                Box::pin(async move {
+                    match method {
+                        "state_getStorage" => {
```

Comment:
It would be quite useful to collect of rpc methods from the Backend to know that we cover the all exposed in that API such as method names changes etc but this works too.

Reply:
I mean a lot of them follow the same structure for legacy backend, but I can re add the tests for other methods too, it's just that they would add more changes by themselves by requiring serializers for expected data

Reply:
yeah, it doesn't work nicely with subscriptions anyway so ignore this